### PR TITLE
Postgres: Sync table row count

### PR DESCRIPTION
**Pull request by qnkhuat**
Milestone 1 of : https://github.com/metabase/metabase/issues/38229, part of https://github.com/metabase/metabase/issues/39176

This epic is about introducing a new fingerprint process that will do a full table scan on the table `select count(distinct(name)) from table` for example.

In order to not blow up DW, we want to limit to only doing this on reasonably small tables (< 1M rows to start with) by syncing the `estimated-row-count` of a table.

### CODE REVIEW

File: resources/migrations/001_update_migrations.yaml
Comment (crisptrutski) [inline resources/migrations/001_update_migrations.yaml:5610]: I'm curious how much thought has been given to the name "properties" - being so generic, are we intentionally leaving it open for ad hoc future use cases? I personally like to error towards narrow and specific, as one can always add further columns. What do you think about calling it something like `statistics`?

File: src/metabase/driver.clj
Comment (crisptrutski) [inline src/metabase/driver.clj:979]: nitpick: "the estimated row count of all tables ..." sounds like a summation over every table, how about:

> Returns a map of [schema, table name] -> the estimated row count, for each table in the current database.

File: src/metabase/driver/postgres.clj
Comment (crisptrutski) [inline src/metabase/driver/postgres.clj:910]: What do you think about using some "as" aliases in the query to make the mapping function read a bit nicer?

e.g. `(fn [{:keys [schema table row_count]}] ...)`

File: src/metabase/driver/sql_jdbc/sync/describe_database.clj
Comment (qnkhuat) [reply src/metabase/driver/sql_jdbc/sync/describe_database.clj:224]: I'm not fond of `join` as it makes me think of join in SQL, wdyt about `hydrate-table-properties`? `add` would be okay to IMO.

File: src/metabase/sync/interface.clj
Comment (qnkhuat) [reply src/metabase/sync/interface.clj:17]: I think it's a bit of a taste, IMO sync interface should all be closed as it forces us to keep the schema up-to-date as it's easy to forget about it :)

File: resources/migrations/001_update_migrations.yaml
Comment (qnkhuat) [reply resources/migrations/001_update_migrations.yaml:5610]: I indeed wanted this to be generic, IMO optional fields like `show_in_getting_started` or `database_require_filter` could fall into this category. 

It actually started as Cam suggested me to put `database_require_filter` into a json column like this in the first place, and I think that's a good idea :)

File: src/metabase/sync/interface.clj
Comment (crisptrutski) [reply src/metabase/sync/interface.clj:17]: In this case it might be annoying if we try to work on an old branch with a new database. I don't think that trumps your reasoning, I was just curious, and in any case also a conservative clam when it comes to typing sensibilities 😄

File: src/metabase/driver/sql_jdbc/sync/describe_database.clj
Comment (crisptrutski) [reply src/metabase/driver/sql_jdbc/sync/describe_database.clj:223]: I'm not super keen on `hydrate`, as we we use that term for toucan models. How about `calculate-table-properties`? Still fine with `add` as well, but thinking now it's nice to indicate that this function "does some work".

File: resources/migrations/001_update_migrations.yaml
Comment (crisptrutski) [reply resources/migrations/001_update_migrations.yaml:5610]: If we explicitly want this to be a wide assortment the name is fine I guess. I'm slightly allergic to generic names but don't have a counter suggestion.

File: src/metabase/driver/sql_jdbc/sync/describe_database.clj
Comment (qnkhuat) [reply src/metabase/driver/sql_jdbc/sync/describe_database.clj:199]: `tables` here is an eduction and I don't want to turn it into a seq yet.

File: src/metabase/driver.clj
Comment (qnkhuat) [reply src/metabase/driver.clj]: [81087df](https://github.com/metabase/metabase/pull/39494/commits/81087df06972202cf006c1f21214e18071bc7f24)

File: src/metabase/driver/sql_jdbc/sync/describe_database.clj
Comment (qnkhuat) [reply src/metabase/driver/sql_jdbc/sync/describe_database.clj]: I'm not sure about `calculate` since to me it means doing some arithmetic calculation, but here we just fetch the row count from DB.

That's said I went with `add-table-properties` here :)

**Comment by crisptrutski** [comment]
I've been burned by these estimates being by off by orders of magnitude. They're typically used for query planning and my experience with them is biased to when I've needed to debug sudden performance issues, tracing back to illogical query plans, tracing back to bogus numbers. My experience was with MySQL, and quite some time ago, and I'm not sure how far the tech has come since then, but I have some lingering apprehension. 

What is the use case for these the counts, and how tolerant will that be of volatility and inaccuracy?

File: src/metabase/driver/sql_jdbc/sync/describe_database.clj
Comment (crisptrutski) [inline src/metabase/driver/sql_jdbc/sync/describe_database.clj:200]: Can we rename this to `:estimated-row-count`? I'm worried about future features not taking these numbers with sufficient salt.

File: test/metabase/driver/postgres_test.clj
Comment (crisptrutski) [inline test/metabase/driver/postgres_test.clj:243]: grammar nit: 

row-count *is* estimated, so the value can't be *known* for sure .. So for tests that *don't* ~~concern~~ require ...

Comment (qnkhuat) [reply test/metabase/driver/postgres_test.clj]: updated in https://github.com/metabase/metabase/pull/39494/commits/1ef0ec58be8a5ae875ec3c0a42cb13d219d6b94e. Thanks!

File: src/metabase/driver/sql_jdbc/sync/describe_database.clj
Comment (qnkhuat) [reply src/metabase/driver/sql_jdbc/sync/describe_database.clj]: yep, renamed in https://github.com/metabase/metabase/pull/39494/commits/1ef0ec58be8a5ae875ec3c0a42cb13d219d6b94e

**Comment by qnkhuat** [comment]
@crisptrutski I've updated the PR description to better clarify the usage of `estimated-row-count`, I'd not say we require this row count to be highly accurate but if it's order of magnitude off then that's no good.

At least in PG, the auto vacuum process runs every minute (by default), so I'd expect it not to be "that" off https://www.postgresql.org/docs/current/routine-vacuuming.html#AUTOVACUUM

File: test/metabase/driver/postgres_test.clj
Comment (crisptrutski) [inline test/metabase/driver/postgres_test.clj]: I'm not sure how precise you expect the stats to be afterward a vacuum, but as far as I know it won't directly update them and the improvements happening will be indirect through updating things like the visibility map. I think you would want to call `ANALYZE`. Even then I believe it does sampling and there are no guarantees at scale.

File: src/metabase/driver/sql_jdbc/sync/describe_database.clj
Comment (crisptrutski) [reply src/metabase/driver/sql_jdbc/sync/describe_database.clj]: Fair enough. For me touching a database is doing work 😄

**Review by crisptrutski — APPROVED** [review]

**Comment by crisptrutski** [comment]
> @crisptrutski I've updated the PR description to better clarify the usage of `estimated-row-count`, I'd not say we require this row count to be highly accurate but if it's order of magnitude off then that's no good.

Large errors will be pretty uncommon. They'll mostly affect tables that are already pretty large, probably outside that limit.

File: resources/migrations/001_update_migrations.yaml
Comment (calherries) [inline resources/migrations/001_update_migrations.yaml:5604]: Remind me, what's the benefit of stuffing this data in a JSON blob over nullable columns? I've tried to make my own list here and I'm not sure it stacks up:

The benefits:
- you can separate fields that are generally required from fields that are specific to certain driver features?
- you don't have to update as many tests

The downsides:
- harder to migrate
- harder to inspect with SQL
- slower to filter/select on

I genuinely feel like I'm missing something though.

File: src/metabase/driver.clj
Comment (calherries) [inline src/metabase/driver.clj:979]: I know we hold all tables in memory in a few places in sync code, but I'm wondering whether we should aim for better and instead return a reducible collection here, so we can choose not to hold everything in memory in the future.

Based on [this question on hosting insights](https://stats.metabase.com/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7ImRhdGFiYXNlIjoyNiwidHlwZSI6InF1ZXJ5IiwicXVlcnkiOnsiYWdncmVnYXRpb24iOltbImNvdW50Il1dLCJicmVha291dCI6W1siZmllbGQiLCJpbnN0YW5jZV9pZCIseyJiYXNlLXR5cGUiOiJ0eXBlL1VVSUQifV0sWyJmaWVsZCIsImRhdGFiYXNlX2lkIix7ImJhc2UtdHlwZSI6InR5cGUvSW50ZWdlciJ9XV0sIm9yZGVyLWJ5IjpbWyJkZXNjIixbImFnZ3JlZ2F0aW9uIiwwXV1dLCJzb3VyY2UtdGFibGUiOiJjYXJkX18xMzQzNCJ9fSwiZGlzcGxheSI6InRhYmxlIiwicGFyYW1ldGVycyI6W10sInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9LCJvcmlnaW5hbF9jYXJkX2lkIjoxMzQzNH0=) we have some pretty large databases out there in the wild. 

I tested the memory footprint of a map similar to the one returned by `schema+table->estimated-row-count` with 1,000,000 tables, and it comes to over 200 MB. I'm pretty sure we'd prefer not to do that.

Here's my test code:
```
(require '[metabase.test :as mt])

(defn gc []
  (dotimes [_ 4] (System/gc)))

(defn used-memory []
  (let [runtime (Runtime/getRuntime)]
    (gc)
    (- (.totalMemory runtime) (.freeMemory runtime))))

(defn mbs [f]
  (let [before (used-memory)
        _ (f)
        after (used-memory)]
    (int (/ (- after before) (* 1024 1024)))))

(mbs (fn []
           (def x (into {} (take 1000000 (map vector (repeatedly (fn [] [(mt/random-name) (mt/random-name)])) (repeat 1000)))))))
=> 232 (MB)
```

Comment (calherries) [reply src/metabase/driver.clj:979]: To be clear I'm not advocating against holding this data in memory now, as we're currently doing that elsewhere so this is not a memory bottleneck. I'm just suggesting to have a path to not holding the data in memory without having to change the driver method implementations.

**Review by calherries — COMMENTED** [review]
This is the second time in recent history where we're adding a new query to save table metadata during sync where we join its results with the results of another query in Clojure. The last time was when we introduced the current user table privileges check. Now it's row counts.

Every time we add a new query it costs time, memory, and extra clojure code to join the data in clojure-land and iterate/reduce over the results. We currently have these queries for table metadata:

1. Table name, schema, description and type
2. Current user select privileges by table
3. Current user create/update/delete privileges by table

This PR will add a fourth query for table row counts. I wonder, when do we bite the bullet and just write a single query for all table metadata per driver?

Switching to do all of this in one big query will never make sense on an individual feature level, but looking at the system as a whole, it will be a lot easier to change and optimize.

One concrete benefit of having one query is this feature would have been easier to implement. You would just add a join to the query. Instead, you've had to create a new driver multimethod which comes packaged with assumptions about memory usage that I think we'll want to break one day.

Another concrete benefit is we can solve memory worries once and for all, instead of having ad-hoc strategies to iterate/page over results per query.

The downside of this proposed approach is that we'd have to maintain the implementation of the logic that JDBC's `getTables` gives us. But my guess is that's not a big cost compared to the benefits. Take a look at the [postgres implementation](https://github.com/davecramer/pgjdbc/blame/2fa24b3a0939c193e79e1ba5f71cb260eae4b148/pgjdbc/src/main/java/org/postgresql/jdbc/PgDatabaseMetaData.java#L1272) of `getTables`: I can't see any changes in the last 5 years that would have required us to update an implementation that meets our needs.

Thoughts? Should we pause this project to refactor first?

**Comment by camsaul** [comment]
We used to sync row count but took it out because it was extremely slow. Putting `select count(distinct(name)) from table` back in is probably going to make things extremely slow again.

File: resources/migrations/001_update_migrations.yaml
Comment (camsaul) [reply resources/migrations/001_update_migrations.yaml:5604]: Agree with Cal. I think this just belongs in a regular column. The old version was a regular column

File: src/metabase/driver.clj
Comment (camsaul) [reply src/metabase/driver.clj:979]: 200 MB is really not ok, we want people (at least theoretically) to be able to run Metabase on Heroku or small EC2 instances. We need to rework this to not hold everything in memory at once. Metabase shouldn't need a gig of RAM to sync a database

File: src/metabase/sync/interface.clj
Comment (camsaul) [reply src/metabase/sync/interface.clj:17]: I prefer leaving maps open, that way drivers can be backwards compatible with older versions if we introduce a new key somewhere.

**Review by camsaul — DISMISSED** [review]
I think Cal is right and the estimated row count should be its own column, its doesn't make sense to introduce another JSON blob for this.

Also please let's just stick with the estimated row count and not introduce any full table scans. Take that out of the epic. It's going to destroy sync performance (already bad) and make everyone mad at us. We took it out for a reason

**Comment by qnkhuat** [comment]
@camsaul this will now be synced to `metabase_table.esimated_row_count`.

> Also please let's just stick with the estimated row count and not introduce any full table scans. Take that out of the epic. It's going to destroy sync performance (already bad) and make everyone mad at us. We took it out for a reason

we(me and Arakaki and Cal) had a chat about this and we agreed to put a hold on that.

File: resources/migrations/001_update_migrations.yaml
Comment (calherries) [inline resources/migrations/001_update_migrations.yaml:6068]: This needs to be a biginteger type to match `pg_stat_user_tables.n_live_tup`. Otherwise it'll fail for tables over ~2 billion rows

File: test/metabase/driver/postgres_test.clj
Comment (calherries) [inline test/metabase/driver/postgres_test.clj:283]: Just for my understanding, why do you not have to execute `VACUUM` here but you do in the [other test](https://github.com/metabase/metabase/pull/39494/files#diff-821f0be02cebbea5f2f5b86239bd9df5e5cb4f671634a761fa6da1c511487125R1429)?

**Review by calherries — DISMISSED** [review]
Sweet, this looks pretty neat now

I left a few minor suggestions, but I'll approve once this is resolved https://github.com/metabase/metabase/pull/39494/files#r1534324338.

File: src/metabase/sync/interface.clj
Comment (qnkhuat) [reply src/metabase/sync/interface.clj]: >  when the driver supports :sync-estimated-row-count. 

I meant to remove that feature since we don't have a partial reason for having it now. maybe we need to let third party drivers know about it but it doesn't have much benefit in syncing it now, do you it still worth mentioning in driver log?

Comment (qnkhuat) [reply src/metabase/sync/interface.clj]: also made this schema a closed map to avoid situations like this.

File: resources/migrations/001_update_migrations.yaml
Comment (qnkhuat) [reply resources/migrations/001_update_migrations.yaml]: good catch! updated [8e78370](https://github.com/metabase/metabase/pull/39494/commits/8e78370cb7edb0b9846a615e587bf9c4a9034635)

File: test/metabase/driver/postgres_test.clj
Comment (qnkhuat) [reply test/metabase/driver/postgres_test.clj:283]: for the `sync-estimated-row-count-test` test I want an accurate number, so executing VACUUM to trigger Postgres to get the exact number. but for these tests we don't need to test the exact number.

File: test/metabase/driver/postgres_test.clj
Comment (qnkhuat) [reply test/metabase/driver/postgres_test.clj]: good point, use `test-data` and moved it to a more generic namespace here sync-estimated-row-count-test

File: resources/migrations/001_update_migrations.yaml
Comment (qnkhuat) [reply resources/migrations/001_update_migrations.yaml]: it's now in its own column `metabase_table.estimated_row_count`

File: src/metabase/sync/sync_metadata/tables.clj
Comment (qnkhuat) [inline src/metabase/sync/sync_metadata/tables.clj:196]: the `i/DatabaseMetadataTable` schema is now a closed one, so it was failing because the table from our metadata includes an `:id`. rather than adding an `:id` to `i/DatabaseMetadataTable`, I think it makes more sense that our-metadata has its own schema here. This is how we do in sync fields as well.

File: src/metabase/sync/interface.clj
Comment (calherries) [reply src/metabase/sync/interface.clj]: As I said on slack I guess we technically don't need a driver feature yet, but when we use the row count we should definitely add the feature to help with testing and optimizations. So up to you. There's not much of a downside if we introduce a driver feature, never use it, then remove it later, so I might just add one now.

Comment (calherries) [reply src/metabase/sync/interface.clj]: > made this schema a closed map

Hmm maybe we want to be careful here, assuming we want to turn on malli schema validation in production again, making this map closed constitutes a breaking change. Driver implementations can currently return whatever keys they want in this map, and if we turn malli schema checking on this will break their sync process. Unlikely, but possible

File: test/metabase/driver/sql_jdbc_test.clj
Comment (calherries) [inline test/metabase/driver/sql_jdbc_test.clj:289]: The previous comment got lost because the code was removed, so here it is
https://github.com/metabase/metabase/pull/39494/files/5e15c627f41d4fa81f89680ab05cae675f745828#r1534344084

I mentioned this should be in a sync test namespace, but this is in a driver test namespace. If we're to take modularization seriously there shouldn't be any sync related code in this namespace.

Maybe a good place to put this test is `metabase.sync.sync-metadata.tables-test`?

It's unfortunate this test uses `(next.jdbc/execute! conn ["VACUUM;"])` which is postgres specific but I think we can worry about that when we get to having multiple drivers support this feature

File: test/metabase/driver/postgres_test.clj
Comment (calherries) [reply test/metabase/driver/postgres_test.clj:283]: got it, thanks

File: test/metabase/driver/postgres_test.clj
Comment (crisptrutski) [inline test/metabase/driver/postgres_test.clj]: I'm a bit worried about this flaking, as there's no guarantee this fixes the statistics. AFAIK `ANALYZE` will be more reliable.

File: src/metabase/api/notify.clj
Comment (crisptrutski) [inline src/metabase/api/notify.clj:62]: It's slightly shorter and slightly more efficient to just check the individual fields.

```suggestion
                               (when (and (= schema_name (:schema table-in-db)))
                                          (= table_name (:name table-in-db)))
```

File: src/metabase/sync/interface.clj
Comment (qnkhuat) [reply src/metabase/sync/interface.clj]: both fair points, I guess we can add the feature now, and also open up the schema again

File: test/metabase/driver/postgres_test.clj
Comment (crisptrutski) [inline test/metabase/driver/postgres_test.clj:262]: nitpick: I prefer 1 key-value per line unless a map is a 1-liner.

File: test/metabase/driver/postgres_test.clj
Comment (crisptrutski) [inline test/metabase/driver/postgres_test.clj:265]: The way we typically use this naming schema I'd expect this to either return a map, or to take a "describe-database" as an argument. How about `describe-tables`?

File: test/metabase/driver/sql_jdbc_test.clj
Comment (crisptrutski) [reply test/metabase/driver/sql_jdbc_test.clj:289]: Perhaps vacuum (or analyze) could be put behind a test multimethod (something like `update-table-statistics!`)

**Review by crisptrutski — COMMENTED** [review]
Looks great, just waiting on Cal's call out.

File: test/metabase/driver/postgres_test.clj
Comment (qnkhuat) [reply test/metabase/driver/postgres_test.clj]: > AFAIK ANALYZE will be more reliable.
interesting, how so?

on flaking: the test-data db is quite small so vacuuming shouldn't take a lot of time. and also no one modify it, so it should be pretty safe IOM.

File: test/metabase/driver/sql_jdbc_test.clj
Comment (qnkhuat) [reply test/metabase/driver/sql_jdbc_test.clj:289]: Moved to metabase.sync.sync-metadata.tables-test

on making vacuum a multimethod: absolutely, but I'll get to it once we add more driver supports :)

**Review by calherries — APPROVED** [review]
LGTM 

you've just got an extra bracket at the end of line 61
https://github.com/metabase/metabase/pull/39494/files#diff-d7645fcaa393ee95545bbf8d6f9fc58a52f375e1ffb1dc6c1a6f8d9ee833efd0R61-R62

File: src/metabase/api/notify.clj
Comment (qnkhuat) [inline src/metabase/api/notify.clj:61]:
```suggestion
                               (when (and (= schema_name (:schema table-in-db))
```

**Pull request merged**
